# Assignment-3.1-MOOC By Rakshit Batra

### 1) Forecasting Accident Type

### Importing all the necessary libraries and the dataset

In [82]:
import numpy as np
import pandas as pd
df = pd.read_csv('/Users/rakshitbatra/Downloads/NHAIAccidentData-1.csv')
df.head()

Date TimeOfAcc     AccLocation  NatureAccident  \
0   9/6/2015  09:00 AM     Ch. 69+5006               2   
1   9/5/2014  07:18 AM          72+500               7   
2  9/10/2014  08:50 AM      20+250 LHS               4   
3   8/9/2015  05:00 AM             NaN               1   
4   8/4/2015  08:23 AM  126.800 KM RHS               2   

  ClassificationOfAccident Causes  RoadFeature  RoadCondition  \
0                        2      2            2              1   
1                        1      2            2              3   
2                        -      3            3              1   
3                        2      2            1              3   
4                        2      2            4              4   

   IntersectionTypeControl  WeatherCondition  \
0                        9                 1   
1                        9                 1   
2                        9                 1   
3                        9                 1   
4                        9                 1   

                          VehicleResponsible  Fatal  Grevious  Minor  Injured  \
0                 One Diesel Truck collapsed      0         2      1        0   
1                            Lorry & Scorpio      2         0      0        0   
2  LCV (KA02 AD 4041) HIT CAR (KA 19 Z 8477)      0         0      0        0   
3                                Lorry Truck      0         1      1        0   
4                                        Car      0         2      1        1   

   NumAnimalsKilled HelpProvidedByAmbulancePatrol  \
0                 0                Petrol Vehicle   
1                 0                     Ambulance   
2                 0                Petrol Vehicle   
3                 0                Petrol Vehicle   
4                 0                Petrol Vehicle   

                                      Remarks  
0                                         NaN  
1                                         NaN  
2  NO PROPERTY DAMAGES. INFORMED KEY PERSONS.  
3                                         NaN  
4              A car upside due to over speed

### Checking for unique elements in ClassificationOfAccident column and counting total occurences against that value.

In [83]:
df['ClassificationOfAccident'].value_counts()

2    25
1    22
3    22
4    14
-     8
Name: ClassificationOfAccident, dtype: int64

### Checking for total unique values in all columns of the dataset

In [84]:
df.nunique()

Date                             65
TimeOfAcc                        54
AccLocation                      77
NatureAccident                    7
ClassificationOfAccident          5
Causes                            7
RoadFeature                       4
RoadCondition                     4
IntersectionTypeControl           4
WeatherCondition                  2
VehicleResponsible               69
Fatal                             6
Grevious                          6
Minor                             5
Injured                           6
NumAnimalsKilled                  2
HelpProvidedByAmbulancePatrol     3
Remarks                          34
dtype: int64

### Duplicating the dataset

In [85]:
df1 = df

### Checking for null ('-') entries in Causes column

In [123]:
for i in range(0,len(df1)):
    if df1.iloc[i]['Causes'] == "-":
        print(i,"Null")

36 Null
54 Null


### Dropping the rows with '-' values in it.

In [124]:
df1.drop(df1[(df1['Causes']=="-")].index, inplace=True)

### Verifying the column once again to make sure that the rows with null/'-' values are deleted.

In [125]:
count = 0
for i in range(0,len(df1)):
    if df1.iloc[i]['Causes'] == "-":
        count+=1
print("Count:",count)

Count: 0


### Checking for null ('-') entries in ClassificationOfAccident column

In [86]:
for i in range(0,len(df1)):
    if df1.iloc[i]['ClassificationOfAccident'] == "-":
        print(i,"Null")

2 Null
11 Null
13 Null
40 Null
48 Null
56 Null
58 Null
86 Null


### Dropping the rows with Null/'-'

In [179]:
df1.drop(df1[(df1['ClassificationOfAccident']=="-")].index, inplace=True)

### Verifying the column once again to make sure that the rows with null/'-' values are deleted.

In [88]:
count = 0
for i in range(0,len(df1)):
    if df1.iloc[i]['ClassificationOfAccident'] == "-":
        count+=1
print("Count:",count)

Count: 0


In [89]:
#Checking for null values in AccLocation column
df1[(df1['AccLocation'].isna())]
#Hence, Dropping those rows
df1.drop(df1[(df1['AccLocation'].isna())].index, inplace=True)

### Replacing + with . in AccLocation's Entries

In [90]:
df1['AccLocation'] = df1['AccLocation'].str.replace('+','.',regex=True)

In [91]:
df1.describe()

NatureAccident  RoadFeature  RoadCondition  IntersectionTypeControl  \
count       81.000000    81.000000      81.000000                81.000000   
mean         3.000000     3.061728       2.629630                 8.407407   
std          1.981161     0.953227       1.400397                 1.954340   
min          1.000000     1.000000       1.000000                 1.000000   
25%          2.000000     2.000000       1.000000                 9.000000   
50%          2.000000     3.000000       3.000000                 9.000000   
75%          3.000000     4.000000       4.000000                 9.000000   
max          7.000000     4.000000       4.000000                 9.000000   

       WeatherCondition      Fatal   Grevious      Minor    Injured  \
count         81.000000  81.000000  81.000000  81.000000  81.000000   
mean           1.024691   0.530864   0.938272   0.604938   0.666667   
std            0.222222   1.215385   1.486150   0.846744   1.303840   
min            1.000000   0.000000   0.000000   0.000000   0.000000   
25%            1.000000   0.000000   0.000000   0.000000   0.000000   
50%            1.000000   0.000000   0.000000   0.000000   0.000000   
75%            1.000000   1.000000   2.000000   1.000000   1.000000   
max            3.000000   8.000000  10.000000   5.000000   5.000000   

       NumAnimalsKilled  
count         81.000000  
mean           0.024691  
std            0.156150  
min            0.000000  
25%            0.000000  
50%            0.000000  
75%            0.000000  
max            1.000000

### Checking for total number of occurences for value 0.

In [92]:
print(type(df1['NumAnimalsKilled'][0]))
count = 0
for i in range(0,len(df1)):
    if df1.iloc[i]['NumAnimalsKilled'] == 0:
        count +=1
print(count)

<class 'numpy.int64'>
79


### Checking for total number of occurences for value > 0.

In [93]:
count = 0
for i in range(0,len(df1)):
    if df1.iloc[i]['NumAnimalsKilled'] > 0:
        count +=1
print(count)

2


### Since, 79 out of 81 entries are 0 which means it will not affect our output hence we will drop those columns.

In [94]:
df1 = df1.drop(['NumAnimalsKilled'], axis=1)

### Checking for unique elements and their occurences in the following columns.

#### Injured

In [95]:
#checking for unique values in injured column and their occurences
arr = df1['Injured'].unique()
list = []
for i in arr:
    list.append(i)
list.sort()
print(list)
for j in list:
    count = 0
    for i in range(0,len(df1)):
        if df1.iloc[i]['Injured'] == j:
            count +=1
    print(j,":",count)

[0, 1, 2, 3, 4, 5]
0 : 60
1 : 4
2 : 8
3 : 5
4 : 1
5 : 3


#### Minor

In [96]:
#checking for unique values in Minor column and their occurences
arr = df1['Minor'].unique()
list = []
for i in arr:
    list.append(i)
list.sort()
print(list)
for j in list:
    count = 0
    for i in range(0,len(df1)):
        if df1.iloc[i]['Minor'] == j:
            count +=1
    print(j,":",count)

[0, 1, 2, 3, 5]
0 : 44
1 : 29
2 : 6
3 : 1
5 : 1


#### Grevious

In [97]:
#checking for unique values in Grevious column and their occurences
arr = df1['Grevious'].unique()
list = []
for i in arr:
    list.append(i)
list.sort()
print(list)
for j in list:
    count = 0
    for i in range(0,len(df1)):
        if df1.iloc[i]['Grevious'] == j:
            count +=1
    print(j,":",count)

[0, 1, 2, 3, 5, 10]
0 : 42
1 : 17
2 : 18
3 : 1
5 : 2
10 : 1


#### Fatal

In [98]:
#checking for unique values in Fatal column and their occurences
arr = df1['Fatal'].unique()
list = []
for i in arr:
    list.append(i)
list.sort()
print(list)
for j in list:
    count = 0
    for i in range(0,len(df1)):
        if df1.iloc[i]['Fatal'] == j:
            count +=1
    print(j,":",count)

[0, 1, 2, 3, 5, 8]
0 : 59
1 : 11
2 : 8
3 : 1
5 : 1
8 : 1


#### HelpProvidedByAmbulancePatrol

In [99]:
#checking for unique values in HelpProvidedByAmbulancePatrol column and their occurences
arr = df1['HelpProvidedByAmbulancePatrol'].unique()
list = []
for i in arr:
    list.append(i)
list.sort()
print(list)
for j in list:
    count = 0
    for i in range(0,len(df1)):
        if df1.iloc[i]['HelpProvidedByAmbulancePatrol'] == j:
            count +=1
    print(j,":",count)

['Ambulance', 'Ambulance/Petrol Vehicle', 'Petrol Vehicle']
Ambulance : 13
Ambulance/Petrol Vehicle : 20
Petrol Vehicle : 48


### Now, Verifying the dataset just to make sure all the necessary changes are there

In [100]:
df1.head()

Date TimeOfAcc     AccLocation  NatureAccident  \
0  9/6/2015  09:00 AM     Ch. 69.5006               2   
1  9/5/2014  07:18 AM          72.500               7   
4  8/4/2015  08:23 AM  126.800 KM RHS               2   
5  8/4/2015  08:23 AM  126.800 KM RHS               2   
6  8/4/2015  08:23 AM  126.800 KM RHS               2   

  ClassificationOfAccident Causes  RoadFeature  RoadCondition  \
0                        2      2            2              1   
1                        1      2            2              3   
4                        2      2            4              4   
5                        3      2            4              4   
6                        4      2            4              4   

   IntersectionTypeControl  WeatherCondition          VehicleResponsible  \
0                        9                 1  One Diesel Truck collapsed   
1                        9                 1             Lorry & Scorpio   
4                        9                 1                         Car   
5                        9                 1                         Car   
6                        9                 1                         Car   

   Fatal  Grevious  Minor  Injured HelpProvidedByAmbulancePatrol  \
0      0         2      1        0                Petrol Vehicle   
1      2         0      0        0                     Ambulance   
4      0         2      1        1                Petrol Vehicle   
5      0         2      1        1                Petrol Vehicle   
6      0         2      1        1                Petrol Vehicle   

                          Remarks  
0                             NaN  
1                             NaN  
4  A car upside due to over speed  
5  A car upside due to over speed  
6  A car upside due to over speed

### Encoding the HelpProvidedByAmbulancePatrol column values

In [126]:
mapping = {'Ambulance': 0, 'Ambulance/Petrol Vehicle': 1, 'Petrol Vehicle':2}
df1 = df1.replace({'HelpProvidedByAmbulancePatrol': mapping})

### Now, verifying the dataset to make sure that encoding is successfully done.

In [127]:
df1.head()

Date TimeOfAcc     AccLocation  NatureAccident  \
0  9/6/2015  09:00 AM     Ch. 69.5006               2   
1  9/5/2014  07:18 AM          72.500               7   
4  8/4/2015  08:23 AM  126.800 KM RHS               2   
5  8/4/2015  08:23 AM  126.800 KM RHS               2   
6  8/4/2015  08:23 AM  126.800 KM RHS               2   

  ClassificationOfAccident Causes  RoadFeature  RoadCondition  \
0                        2      2            2              1   
1                        1      2            2              3   
4                        2      2            4              4   
5                        3      2            4              4   
6                        4      2            4              4   

   IntersectionTypeControl  WeatherCondition          VehicleResponsible  \
0                        9                 1  One Diesel Truck collapsed   
1                        9                 1             Lorry & Scorpio   
4                        9                 1                         Car   
5                        9                 1                         Car   
6                        9                 1                         Car   

   Fatal  Grevious  Minor  Injured  HelpProvidedByAmbulancePatrol  \
0      0         2      1        0                              2   
1      2         0      0        0                              0   
4      0         2      1        1                              2   
5      0         2      1        1                              2   
6      0         2      1        1                              2   

                          Remarks  
0                             NaN  
1                             NaN  
4  A car upside due to over speed  
5  A car upside due to over speed  
6  A car upside due to over speed

### Checking for total number of rows with Dates ending with year = 2014

In [128]:
li = []
for i in df1['Date']:
    if '2014' in i:
        li.append(i)
print(len(li))


30


### Separating those rows into another dataset with the name 'df2014'

In [129]:
df2014=df1.loc[df1['Date'].isin(li)]

### Verifying its length to make sure that all rows are being added to the new dataset

In [130]:
print(len(df2014))
df2014.head()

30


Date TimeOfAcc AccLocation  NatureAccident ClassificationOfAccident  \
1    9/5/2014  07:18 AM      72.500               7                        1   
14   5/6/2014  01:25 PM      80.000               7                        3   
18  5/12/2014  09:00 AM  91.800 LHS               2                        1   
20   4/5/2014  05:15 AM   60.200RHS               1                        1   
22  30/9/2014  10:30 PM  14.835 LHS               7                        2   

   Causes  RoadFeature  RoadCondition  IntersectionTypeControl  \
1       2            2              3                        9   
14      3            2              1                        3   
18      3            2              1                        1   
20      4            2              4                        9   
22      5            3              1                        9   

    WeatherCondition                                 VehicleResponsible  \
1                  1                                    Lorry & Scorpio   
14                 1                    Lorry RJ19 GA2487 vs Pedestrian   
18                 1                                        Motor Cycle   
20                 1                                Two Wheeler & Lorry   
22                 1  LCV(KA 01 AA 1583) HIT KSRTC BUS(KA 19 F 2831)...   

    Fatal  Grevious  Minor  Injured  HelpProvidedByAmbulancePatrol  \
1       2         0      0        0                              0   
14      0         0      1        0                              2   
18      1         2      1        0                              0   
20      2         2      0        0                              0   
22      0         1      0        0                              1   

                                              Remarks  
1                                                 NaN  
14                                                NaN  
18                                                Nil  
20                                                NaN  
22  INJURED PERSON SHIFTED TO HARSH HOSPITAL, NO P...

### Checking for total number of rows with Dates ending with the name 'df2015'

In [131]:
li = []
for i in df1['Date']:
    if '2015' in i:
        li.append(i)

In [132]:
df2015=df1.loc[df1['Date'].isin(li)]
print(len(df2015))
df2015.head()

45


Date TimeOfAcc     AccLocation  NatureAccident  \
0  9/6/2015  09:00 AM     Ch. 69.5006               2   
4  8/4/2015  08:23 AM  126.800 KM RHS               2   
5  8/4/2015  08:23 AM  126.800 KM RHS               2   
6  8/4/2015  08:23 AM  126.800 KM RHS               2   
7  7/4/2015  08:30 AM  160.110 KM RHS               2   

  ClassificationOfAccident Causes  RoadFeature  RoadCondition  \
0                        2      2            2              1   
4                        2      2            4              4   
5                        3      2            4              4   
6                        4      2            4              4   
7                        3      2            4              4   

   IntersectionTypeControl  WeatherCondition          VehicleResponsible  \
0                        9                 1  One Diesel Truck collapsed   
4                        9                 1                         Car   
5                        9                 1                         Car   
6                        9                 1                         Car   
7                        9                 1                         Car   

   Fatal  Grevious  Minor  Injured  HelpProvidedByAmbulancePatrol  \
0      0         2      1        0                              2   
4      0         2      1        1                              2   
5      0         2      1        1                              2   
6      0         2      1        1                              2   
7      0         0      1        0                              2   

                          Remarks  
0                             NaN  
4  A car upside due to over speed  
5  A car upside due to over speed  
6  A car upside due to over speed  
7      accident due to over speed

### Dropping unnecessary columns such as Date, TimeOfAcc, AccLocation, VehicleResponsible, Remarks as these are of string type which will create an error when passed into ML model.

In [133]:
#Now, dropping all unnecessary columns from the 2014 dataset
df2014 = df2014.drop(['Date','TimeOfAcc','AccLocation','VehicleResponsible','Remarks'], axis=1)

In [134]:
#Now, 2014 Dataset is ready for training.
df2014.head()

NatureAccident ClassificationOfAccident Causes  RoadFeature  \
1                7                        1      2            2   
14               7                        3      3            2   
18               2                        1      3            2   
20               1                        1      4            2   
22               7                        2      5            3   

    RoadCondition  IntersectionTypeControl  WeatherCondition  Fatal  Grevious  \
1               3                        9                 1      2         0   
14              1                        3                 1      0         0   
18              1                        1                 1      1         2   
20              4                        9                 1      2         2   
22              1                        9                 1      0         1   

    Minor  Injured  HelpProvidedByAmbulancePatrol  
1       0        0                              0  
14      1        0                              2  
18      1        0                              0  
20      0        0                              0  
22      0        0                              1

In [135]:
#Now, dropping all unnecessary columns from the 2015 dataset and dropping ClassificationOfAccident for prediction.
df2015 = df2015.drop(['Date','TimeOfAcc','AccLocation','VehicleResponsible','Remarks', 'ClassificationOfAccident'], axis=1)

In [136]:
#Now, 2015 Dataset is ready for prediction.
df2015.head()

NatureAccident Causes  RoadFeature  RoadCondition  IntersectionTypeControl  \
0               2      2            2              1                        9   
4               2      2            4              4                        9   
5               2      2            4              4                        9   
6               2      2            4              4                        9   
7               2      2            4              4                        9   

   WeatherCondition  Fatal  Grevious  Minor  Injured  \
0                 1      0         2      1        0   
4                 1      0         2      1        1   
5                 1      0         2      1        1   
6                 1      0         2      1        1   
7                 1      0         0      1        0   

   HelpProvidedByAmbulancePatrol  
0                              2  
4                              2  
5                              2  
6                              2  
7                              2

### Splitting, Importing and Initializing our Model

In [137]:
y = df2014["ClassificationOfAccident"]
x = df2014.drop(["ClassificationOfAccident"], axis=1)

In [138]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
logisticRegr = LogisticRegression(solver='lbfgs', max_iter=200)

In [168]:
X_train,X_test,y_train,y_test=train_test_split(x,y,test_size=1,train_size=.35,random_state=1)

In [169]:
print(X_test)
print(X_train)
print(y_train)
print(y_test)

    NatureAccident Causes  RoadFeature  RoadCondition  \
51               7      2            2              1   

    IntersectionTypeControl  WeatherCondition  Fatal  Grevious  Minor  \
51                        9                 1      0         2      0   

    Injured  HelpProvidedByAmbulancePatrol  
51        0                              1  
    NatureAccident Causes  RoadFeature  RoadCondition  \
66               5      7            3              1   
33               6      3            3              1   
60               7      2            2              3   
37               4      4            3              1   
62               7      2            3              1   
83               4      4            3              1   
20               1      4            2              4   
81               3      3            3              1   
67               4      4            3              1   
72               2      2            3              1   

    IntersectionType

### Feeding our dataset with training values

In [ ]:
classifier= LogisticRegression(random_state=7)  
classifier.fit(X_train,y_train)

### Predicting ClassificationOfAccident for df2015 dataset

In [171]:
y_pred= classifier.predict(df2015)
print(y_pred)

['2' '2' '2' '2' '3' '3' '3' '3' '3' '3' '3' '2' '2' '3' '3' '3' '2' '2'
 '2' '3' '1' '2' '3' '2' '3' '2' '3' '2' '2' '3' '3' '3' '2' '1' '1' '3'
 '3' '2' '2' '2' '3' '1' '1' '3' '3']


### Training Score for our Classifier

In [172]:
classifier.score(X_train, y_train)

0.9

### Testing Score for our Classifier

In [173]:
classifier.score(X_test, y_test)

1.0

### Adding a new column with the predicted values

In [174]:
df2015['Predicted ClassificationOfAccident'] = y_pred


In [175]:
df2015.head()

NatureAccident Causes  RoadFeature  RoadCondition  IntersectionTypeControl  \
0               2      2            2              1                        9   
4               2      2            4              4                        9   
5               2      2            4              4                        9   
6               2      2            4              4                        9   
7               2      2            4              4                        9   

   WeatherCondition  Fatal  Grevious  Minor  Injured  \
0                 1      0         2      1        0   
4                 1      0         2      1        1   
5                 1      0         2      1        1   
6                 1      0         2      1        1   
7                 1      0         0      1        0   

   HelpProvidedByAmbulancePatrol Predicted ClassificationOfAccident  
0                              2                                  2  
4                              2                                  2  
5                              2                                  2  
6                              2                                  2  
7                              2                                  3

### 2) Checking for the Location with maximum number of Accidents

In [177]:
df1['AccLocation'].describe()

count                 79
unique                69
top       126.800 KM RHS
freq                   3
Name: AccLocation, dtype: object

#### Hence, 126.800 KM RHS has maximum number of accidents recorded

### 3) Checking for features of  ClassificationOfAccident column

In [178]:
df1['ClassificationOfAccident'].describe()

count     79
unique     4
top        2
freq      23
Name: ClassificationOfAccident, dtype: object